pyfarm/jobtypes/core/internals.py

```diff
@@ -607,13 +609,30 @@ def result_callback(task, log_path, deferred, response):
                             log_path, task["id"])
                 deferred.callback(None)
 
-        def error_callback(task, log_path, deferred, failure_reason):
-            delay = http_retry_delay()
-            logger.error(
-                "Error while registering logfile %s for task %s on master: "
-                "\"%s\", retrying in %s seconds.",
-                log_path, task["id"], failure_reason, delay)
-            post_logfile(task, log_path, post_deferred=deferred, delay=delay)
+        def error_callback(task, log_path, deferred, num_retry_errors,
+                           failure_reason):
+            if num_retry_errors > config["broken_connection_max_retry"]:
+                logger.error(
+                    "Error while registering logfile %s for task %s on master. "
+                    "Maximum number of retries reached. Not retrying the "
+                    "request.", log_path, task["id"])
+                deferred.errback
+            else:
+                if (failure_reason.type in
+                    (ResponseNeverReceived, RequestTransmissionFailed)):
+                    logger.debug(
+                        "Error while registering logfile %s for task %s on "
+                        "master: %s, retrying immediately",
+                        log_path, task["id"], failure_reason.type.__name__)
+                    post_logfile(task, log_path, post_deferred=deferred)
+                else:
+                    delay = http_retry_delay()
+                    logger.error(
+                        "Error while registering logfile %s for task %s on "
+                        "master: \"%s\", retrying in %s seconds.",
```

Comment:
Use `%r` so you don't have to escape the quotes.  It will also render properly if it happens to not be a string for some reason.